sn_interface/src/network_knowledge/section_peers.rs

```diff
@@ -21,47 +19,71 @@ const ELDER_CHURN_EVENTS_TO_PRUNE_ARCHIVE: usize = 5;
 const ELDER_CHURN_EVENTS_TO_PRUNE_ARCHIVE: usize = 3;
 
 /// Container for storing information about (current and archived) members of our section.
-#[derive(Clone, Default, Debug)]
+#[derive(Clone, Default, Debug, Eq, PartialEq, Ord, PartialOrd)]
 pub(super) struct SectionPeers {
-    members: BTreeMap<XorName, SectionSigned<NodeState>>,
-    archive: BTreeMap<XorName, SectionSigned<NodeState>>,
+    members: BTreeMap<XorName, Decision<NodeState>>,
+    archive: BTreeMap<XorName, Decision<NodeState>>,
 }
 
 impl SectionPeers {
     /// Returns set of current members, i.e. those with state == `Joined`.
-    pub(super) fn members(&self) -> BTreeSet<SectionSigned<NodeState>> {
+    pub(super) fn members(&self) -> BTreeSet<NodeState> {
+        // TODO: ensure Decision has no Joined and Left combined within one
+        let mut node_state_list = BTreeSet::new();
+        for decision in self.members.values() {
+            for node_state in decision.proposals.keys() {
+                let _ = node_state_list.insert(node_state.clone());
+            }
+        }
+        node_state_list
+    }
+
+    /// Returns set of current members, i.e. those with state == `Joined`.
+    /// with the correspondent Decision info
+    pub(super) fn section_members_with_decision(&self) -> BTreeSet<Decision<NodeState>> {
+        // TODO: ensure Decision has no Joined and Left combined within one
         self.members.values().cloned().collect()
     }
 
     /// Returns set of archived members, i.e those that've left our section
-    pub(super) fn archived_members(&self) -> BTreeSet<SectionSigned<NodeState>> {
-        self.archive.values().cloned().collect()
+    pub(super) fn archived_members(&self) -> BTreeSet<NodeState> {
+        // TODO: ensure Decision has no Joined and Left combined within one
+        let mut node_state_list = BTreeSet::new();
+        for decision in self.archive.values() {
+            for node_state in decision.proposals.keys() {
```

Comment:
Again, just add a filter here and we can drop the TODO's
```suggestion
            for node_state in decision.proposals.keys().filter(|node| node.state != MembershipState::Joined) {
```

Reply:
Same comment as above https://github.com/maidsafe/safe_network/pull/2089#discussion_r1107092806